File: Programmers/level 2/괄호변환.js
function solution(p) {
    var answer = '';
    let u='';
    let v='';
    //u, v로 분리하는 과정
    const splitBrace=(p)=>{
        let right=0;
        let left=0;
        u='';
        while (true){        
            let popped=p[0];
            p=p.slice(1);
            u+=popped;
            if (popped==='(') left++;
            else if (popped===')') right++;
            if (left===right) break;
        }
        v=p;
    }
    //균형잡힌 괄호 문자열인지 판별
    const rightBrace=(p)=>{
        let stack=[];
        for (let i=0; i<p.length; i++){
            if (p[i]==="(") stack.push([i]);
            else if (p[i]===")") stack.pop();
        }
        if (stack.length===0) return true;
        else return false;
    }
    //4-4과정
    const reverseBracket=(p)=>{
        return p.slice(1).slice(0,-1).replace( /\(|\)/g , i=>i==="(" ? ")" : "(" );
    }
    const recursive=(p)=>{
        if (p.length===0) return p;
        splitBrace(p);
        if (rightBrace(u)){
             return u+recursive(v);
        }
        else{
            let reverse=reverseBracket(u);
            let temp="("+recursive(v)+")"+reverse;
            return temp;
        }
    }
    answer=recursive(p);
    return answer;
}

File: Programmers/level 2/괄호 변환.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

int splitBrace(string str){
    int left=0;
    int right=0;
    int idx=0;
    while (true){
        if (str[idx]=='(') left++;
    	else right++;
        if (left==right) break;
        idx++;
    }
    return idx+1;
}

bool isCorrect(string str){
    stack<char>s;
    for (int i=0; i<str.length(); i++) {
        if (str[i]=='(') s.push(str[i]);
        else {
            if (s.empty()) return false;
            s.pop();
        }
    }
    if (!s.empty()) return false;
    return true;
}

string reverseBrace(string str) {
    string temp="";
    for (int i=0; i<str.length(); i++) {
        if (str[i]=='(') temp+=')';
        else temp+='(';
    }
    return temp;
}

string recursive (string str) {
    if (str=="") return "";
    
    string u=str.substr(0, splitBrace(str));
    string v=str.substr(splitBrace(str));
    string temp="";
    
    if (isCorrect(u)) {
        temp+=u;
        temp+=recursive(v);
        return temp;
    }
    else {
        temp+='(';
        temp+=recursive(v);
        temp+=')';
        u.erase(u.begin());
        u.erase(u.end()-1);
        temp+=reverseBrace(u);
        return temp;
    }
}

string solution(string p) {
    string answer = "";
    answer=recursive(p);
    return answer;
}
